### DISCUSSION TITLE
storage update and delete policies

### DISCUSSION THREAD

**Discussion by EdgarsLv** [supabase/supabase#5786] [Questions] [answered]
How exactly this works? Because it doesn't works in any way.

![image](https://user-images.githubusercontent.com/56837461/156640682-48ced851-33cb-4b24-bb35-bc1ad34a3fa4.png)

So this is path to old image i want to update/replace? And new file. But then whats the name of new file?

![image](https://user-images.githubusercontent.com/56837461/156640903-3493dd66-8feb-49ef-a3a3-51acbe626500.png)

Whenever i try to update i get this error.

![image](https://user-images.githubusercontent.com/56837461/156641348-579f9400-75bf-445a-bb68-79e1a2e9d789.png)

I can upload and that works fine, but i want only 1 image in folder and update/replace it. No matter what and how i try i just can't update.

**Comment by GaryAustin1** [comment]
Update will replace the old file data with the new file data and not change the name.
Sounds like your policies are not setup for update.
You will need a policy for update (using) and select(using):
https://supabase.com/docs/reference/javascript/storage-from-update
Not sure how you inserted the doc (UI interface or api) but for api you would have needed a policy for insert (with check).

These are independent of your bucket being public.  That just allows access to a public url to read the file.

**Reply by EdgarsLv** [reply]
I have update, insert, select policies. (same as default for avatar bucket)

Don't have delete policy so was excepting to get at least error for removing, but no error and no file removed.

![image](https://user-images.githubusercontent.com/56837461/156648705-c8b4a346-337a-488c-8d71-b941d7d43c86.png)
![image](https://user-images.githubusercontent.com/56837461/156648779-4a6a6c2c-8817-4141-8b3d-16b4dc7e1162.png)

**Reply by EdgarsLv** [reply]
> Update will replace the old file data with the new file data and not change the name.

This is what i was expecting to happen.

**Comment by GaryAustin1** [comment]
Delete without policy, or not meeting policy, will just return an empty array.
I'd try setting your policy to just ((bucket_id = 'userimages') AND (role() = 'authenticated'))and see if that works, Then if not set to 'anon'.  If anon works then your user is not logged in most likely when doing the storage call.  If authenticated works then your policy is likely incorrect.

**Comment by EdgarsLv** [comment, marked as answer]
updated all policies to ((bucket_id = 'userimages') AND (role() = 'anon'))
Cleared cache and site date. Relogged.

Still same error for updating: 'resource was not found',

For deleting response empty array.
Delete and update doesn't works. Just upload.

![image](https://user-images.githubusercontent.com/56837461/156654776-711ad611-f978-46bc-ae7f-6ec8c71cec13.png)

![image](https://user-images.githubusercontent.com/56837461/156655197-716a8b5b-9af6-4e90-b37e-1ef0f50b5cfb.png)

**Reply by GaryAustin1** [reply]
Can you show all your policies including select?  Insert does not require select policy, but update and remove do.

**Reply by EdgarsLv** [reply]
Not sure where i can fully see policy, but they are:

![image](https://user-images.githubusercontent.com/56837461/156656889-7730480b-b2ab-46a2-9882-103f8e204345.png)

plus added:

![image](https://user-images.githubusercontent.com/56837461/156656960-53de26db-d69f-4733-ab6d-bcb3eb67275b.png)

**Reply by GaryAustin1** [reply]
So your policies are a bit out of whack....
Select just needs USING
Insert just needs WITH CHECK
Update needs USING and can optionally have WITH CHECK (but not needed unless different)
Delete needs USING
*** USING looks at existing record, WITH CHECK looks at new record

Then the supabase call for update and remove also requires the select policy work (which it should in your case)

**Reply by EdgarsLv** [reply]
Updated policies and now works update and delete! 

And i wasted many hours trying to figure out why doesn't works...

Thank You very much!

**Reply by GaryAustin1** [reply]
Note that when you delete/replace, the file will still be around for 3600 seconds or cache setting after your operation on browsers and cdn.

**Reply by iamthe-observer** [reply]
> Updated policies and now works update and delete!
> 
> And i wasted many hours trying to figure out why doesn't works...
> 
> Thank You very much!


yh so please im kinda a newbie in sql and polices and im having the same issue as you did so could you by any chance show the policy code so i can have a better understanding of how it works?

**Reply by abhishekmg** [reply]
```
create policy "select" on storage.objects
  for select using (bucket_id = <id>);
  
  create policy "delete" on storage.objects
  for delete using (bucket_id = <id>);
  ```
  
  these are the policies I added to make delete work

**Reply by iamthe-observer** [reply]
I mean... lmao. It's been 4 months and I kinda understand now. But thanks for answering anyway. ✨

**Reply by EmmanuelHra601** [reply]
Hello how are you, my problem is that the image is not updated, it does not give me any error, the image changes in the bucket but when I reload the bucket the previous image reappears. Could you pass me the policies you implemented please?

**Reply by ricardosikic** [reply]
Hi, Did he find the solution for updating the file name and file data as well?

**Reply by iamthe-observer** [reply]
> Hello how are you, my problem is that the image is not updated, it does not give me any error, the image changes in the bucket but when I reload the bucket the previous image reappears. Could you pass me the policies you implemented please?

So I think the way supabase works is that it has a cache system for images you deleted, meaning the pics are still there for a time limit after deletion. But you could change the cache time limit in js/ts.

**Reply by MKirovBG** [reply]
I've spent whole day trying to resolve this issue. Thank you @abhishekmg

**Comment by ABSanthosh** [comment]
In case anyone stumbles on this question, there's already an answer but its not very clear. If you're lazy like me, here's what worked for me.
> So your policies are a bit out of whack.... Select just needs USING Insert just needs WITH CHECK Update needs USING and can optionally have WITH CHECK (but not needed unless different) Delete needs USING *** USING looks at existing record, WITH CHECK looks at new record
> 
> Then the supabase call for update and remove also requires the select policy work (which it should in your case)

If you're lazy like me, here's what worked for me.

```js
const name = "equipments" // your bucket name
const policies = {
select: `
	create policy "select_equipment_image"
	on storage.objects for select
	to authenticated
	using (bucket_id = '${name}');`,
insert: `
	create policy "insert_equipment_image"
	on storage.objects for insert
	to authenticated
	with check (bucket_id = '${name}');`,
update: `
	create policy "update_equipment_image"
	on storage.objects for update
	to authenticated
	using (bucket_id = '${name}');`,
delete: `
	create policy "delete_equipment_image"
	on storage.objects for delete
	to authenticated
	using (bucket_id = '${name}');`
};
```
